PBI-19 Create Test Flow (Authorization)

johaneschristian opened the pull request:
Add an authorization logic to prevent company in using assessment tools that are not associated to them.

johaneschristian committed:
[RED] create test for email with same case functionality

In the previous implementation, there is a hidden bug as email with same
characters but different case can be used to register 2 different
accounts. This commit tests the validation and controller method that
serves the functionality for COMPANY.

johaneschristian committed:
[GREEN] implement functionalities to prevent duplicated email

Store email in its lowercase form and perform validation to the
normalized version of email.

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-01-create_assignment

johaneschristian committed:
[RED] create stub and test for serve_get_user_info

johaneschristian committed:
[GREEN] implement serve_get_user_info functionalities

The function will be used to aid protecting routes at the front end. It
will return the data of the authenticated user.

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-07-user_info

johaneschristian committed:
[RED] create model for test flow, test and stub for utility functions

Create TestFlow and TestFlowTool models for TestFlow creation. The
serializers were also created using the default
serializer.ModelSerializer approach. As TestFlowTool will deals heavily
with AssessmentTool and such distinction between subclasses of
AssessmentTool is required, AssessmentTool now inherits from Polymorphic
model. This commit also creates stub and test for utility function
needed for test flow creation.

johaneschristian committed:
[CHORES] add django-polymorphic to requirements.txt

johaneschristian committed:
[GREEN] implement utility functions for creating test flow

johaneschristian committed:
[RED] create test and stub for adding tool to test flow

johaneschristian committed:
[GREEN] implement add_tool function of TestFlow

johaneschristian committed:
[RED] create stub and test for test flow registration validator

johaneschristian committed:
[GREEN] implement validate_test_flow_registration

johaneschristian committed:
[RED] create test for convert_assessment_tool_id_to_assessment_tool

One of the request attribute is a list of assessment tool id and release
time. This function will be used to convert each of the id to its
respective assessment tool and the release time to be a time object.

johaneschristian committed:
[GREEN] implement convert_assessment_tool_id_to_assessment_tool

johaneschristian committed:
[RED] create tests for save_test_flow_to_database

johaneschristian committed:
[GREEN] implement save_test_flow_to_database

johaneschristian committed:
[RED] add end to end tests for test flow creation

johaneschristian committed:
[GREEN] implement end-point to create test flow

Allow company or assessor to create test flow by submitting request
according to the structure specified.

johaneschristian committed:
[REFACTOR] use InvalidTestFlowRegistration for test flow exception

Earlier, InvalidRequestException was used. To aid better error handling
and to ensure that error handling become more apparent, a more specific
class is introduced. To ensure backward compatibility, the exception
still inherits from InvalidRequestExceptio use
InvalidTestFlowRegistration for test flow exception

Earlier, InvalidRequestException was used. To aid better error handling
and to ensure that error handling become more apparent, a more specific
class is introduced. To ensure backward compatibility, the exception
still inherits from InvalidRequestException.

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-19-create_test_flow

johaneschristian committed:
[RED] create stub and test for assessment tool authorization

Previously, company can use any assessment tool registered in the
database. This create stubs for the necessary functions that prevents that
from happening by ensuring that only tools associated with the company
are retrieved in the previously 'get_tool_by_id' function.

johaneschristian committed:
[GREEN] implement get_tool_of_company_from_id

whenwe177 commented on assessment/services/test_flow.py:
```diff
@@ -20,18 +21,18 @@ def validate_test_flow_registration(request_data: dict):
+            tool = utils.get_tool_of_company_from_id(request_tool_data.get('tool_id'), company)
             release_time = utils.get_time_from_date_time_string(request_tool_data.get('release_time'))
             tool_data = {
                 'tool': tool,
```
Hello, Jo. If I may ask, does the release time means the time the assessees are notified about the assignment? I guess if that's the case I believe adding a new `start_time` might help, which signifies what time the attempt will be opened to the assessees.

johaneschristian replied:
Hello, Alex. I have accomodated that in my latest commit. Thanks for your suggestion.

whenwe177 replied:
Hello, Jo. Thanks for applying this change. I have also created the corresponding component to apply this in the frontend.

johaneschristian committed:
[RED] add test for get_time_from_date_time for None

johaneschristian committed:
[GREEN] implement get_time_from_date_time_string to handle None

johaneschristian committed:
[RED] create tests for adding start_working_time in TestFlow

start_working_time field indicates when will the assessee can first
start doing the activity. This commit create the necessary tests to
evaluate the to be made changes on services and model methods.

johaneschristian committed:
[GREEN] implement functionalities to accomodate start_working_time

johaneschristian committed:
[REFACTOR] change RestrictedAccessException code from 401 to 403
